pkg/operator/ceph/client/controller.go

```diff
@@ -24,295 +24,346 @@ import (
-		return
+		logger.Errorf("failed to reconcile %v", err)
 	}
 
-	err = createClient(c.context, client)
+	return reconcileResponse, err
+}
+
+func (r *ReconcileCephClient) reconcile(request reconcile.Request) (reconcile.Result, error) {
+	// Fetch the CephClient instance
+	cephClient := &cephv1.CephClient{}
+	err := r.client.Get(context.TODO(), request.NamespacedName, cephClient)
 	if err != nil {
-		logger.Errorf("failed to create client %q. %v", client.ObjectMeta.Name, err)
+		if kerrors.IsNotFound(err) {
+			logger.Debug("cephClient resource not found. Ignoring since object must be deleted.")
+			return reconcile.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		return reconcile.Result{}, errors.Wrap(err, "failed to get cephClient")
 	}
-}
 
-func genClientEntity(p *cephv1.CephClient, context *clusterd.Context) (clientEntity string, caps []string, err error) {
-	clientEntity = fmt.Sprintf("client.%s", p.Name)
-	caps = []string{}
+	// Set a finalizer so we can do cleanup before the object goes away
+	err = opcontroller.AddFinalizerIfNotPresent(r.client, cephClient)
+	if err != nil {
+		return reconcile.Result{}, errors.Wrap(err, "failed to add finalizer")
+	}
+
+	// The CR was just created, initializing status fields
+	if cephClient.Status == nil {
+		updateStatus(r.client, request.NamespacedName, cephv1.ConditionProgressing)
+	}
+
+	// Make sure a CephCluster is present otherwise do nothing
+	_, isReadyToReconcile, cephClusterExists, reconcileResponse := opcontroller.IsReadyToReconcile(r.client, r.context, request.NamespacedName, controllerName)
+	if !isReadyToReconcile {
+		// This handles the case where the Ceph Cluster is gone and we want to delete that CR
+		// We skip the deletePool() function since everything is gone already
+		//
+		// Also, only remove the finalizer if the CephCluster is gone
+		// If not, we should wait for it to be ready
+		// This handles the case where the operator is not ready to accept Ceph command but the cluster exists
+		if !cephClient.GetDeletionTimestamp().IsZero() && !cephClusterExists {
+			// Remove finalizer
+			err = opcontroller.RemoveFinalizer(r.client, cephClient)
+			if err != nil {
+				return opcontroller.ImmediateRetryResult, errors.Wrap(err, "failed to remove finalizer")
+			}
 
-	// validate the client settings
-	if err := ValidateClient(context, p); err != nil {
-		return clientEntity, caps, errors.Wrapf(err, "invalid client %q arguments", p.Name)
+			// Return and do not requeue. Successful deletion.
+			return reconcile.Result{}, nil
+		}
+		return reconcileResponse, nil
 	}
 
-	for name, cap := range p.Spec.Caps {
-		caps = append(caps, name, cap)
+	// Populate clusterInfo during each reconcile
+	r.clusterInfo, _, _, err = mon.LoadClusterInfo(r.context, request.NamespacedName.Namespace)
+	if err != nil {
+		return reconcile.Result{}, errors.Wrap(err, "failed to populate cluster info")
 	}
 
-	return clientEntity, caps, nil
-}
+	// DELETE: the CR was deleted
```

Comment:
```suggestion
	// the CR was deleted
```

Reply:
It's the same pattern (line) in the other controllers :-)

Reply:
that's fine, not sure why it stood out this time :)